### DISCUSSION TITLE
Working with CR for version 6

### DISCUSSION THREAD

**Discussion by TheCodeAddiction** [fabric8io/kubernetes-client#4393] [Q&A]
# Intro
I am trying to make an k8 operator that is going to watch and work with custom resources. I am finding a lot of documentation and guides on how to do this, before version 6. I am currently struggling to find up-to date documentation for version 6.

# Prerequisite
1. I have applied my CRD to my k8 cluster.
2. I am running version 6.0.0 of fabric8 kubernetes-client, kubernetes-model and kubernetes-model-common

# Question
I have created a KubernetesClient, using the  `KubernetesClientBuilder().build();` and I am able to watch changes to a specific pod, using the *Watch* class. `Watch watch = client.pods().withName(podName).watch(new Watcher<Pod>(){...}`

Now I want to watch a specific CR. As I understand I need to use _[Generic Resource API](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#resource-api)_? The example code shows how to get a POD in a  namespace `Pod pod = client.resource(pod1).inNamespace("default").get();`. However, I am having a couple of issues
1. It does not show what the argument _pod1_ is. Is it a string? an object? I can't find any documentation about the _client_, which I assume is a `KubernetesClient`, it should be located [in the java doc](https://javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html), but it isn't. 
2. Since I can't find any specific documentation, I am very unsure about how to, or if I can, use this on a CR. 

I have been looking around on the docs here on github, but I am struggling to find any specific information on how to work with CR and I can't find any in-depth _documentation_ about, for example, the KuberentesClient class and it's methods etc. 
This might be a "me" problem, but I have googled in and out for a long time and read everywhere I can. Most online "tutorials" use an older version of fabric8. I have read the [MIGRATION-v6.md](https://github.com/fabric8io/kubernetes-client/blob/master/doc/MIGRATION-v6.md) document and tried to convert old fabric8 code to the newer version, but I am struggling to find in-depth documentation. It seems that if I had worked with fabric8, the migration guide would be amazing, but since I haven't I am a bit stuck. 

# TLDR:
1. How do I watch, GET and Create a CR with fabric8 version 6+
2. Is there any in-depth documentation like [java doc](https://javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) that is up-to date? I can't find any information on, for example: _KubernetesClient_ in the [java doc](https://javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html)

**Comment by rohanKanojia** [comment]
Have you checked these?

https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#resource-typed-api
https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#resource-typeless-api

**Reply by TheCodeAddiction** [reply]
I have no clue how I missed that, but this is great for creating CR!

Is there any in-depth documentation, like java doc, for the KuberentesClient class? I see the example code and can easily make it work, but I would love to read some specification without having to dive into the source code. Is that something that I have also missed?

**Comment by andreaTP** [comment]
@TheCodeAddiction as a reference exists a dedicated project to simplify writing operators in Java:
https://github.com/java-operator-sdk/java-operator-sdk

**Reply by TheCodeAddiction** [reply]
Yes, I know about Operator-SDK :) But we are testing out fabric8 at the place I work at currently, so Operator-SDK is no go for now.

**Comment by TheCodeAddiction** [comment]
@rohanKanojia I am trying to _create_ a CR, however the documentation [you linked above](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#resource-typed-api) is:
1. Deprecated 
2. A bit confusing

### Detail
According to the  [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#resource-typed-api), to create a CR, you do: `cronTabClient.inNamespace("default").create(cronTab1);` 

The issue is, the `create()` method is marked as deprecated. So is every other method in the _ItemWritableOperation_ interface, what am I suppose to use instead? 

Also, what is the `cronTab1` argument? I assume it is an instance of the CronTab class? However, how do I create it correctly? Do I need to create a CronTabSpec and CronTabStatus object and do `cronTab1.setSpec` and `cronTab1.setStatus`? Do I need to create a metadata object and use `cronTab1.setMetadata()`? etc.

**Reply by rohanKanojia** [reply]
you're supposed to use `resource()`
```
cronTabClient.inNamespace("default").resource(cronTab1).create();
```

Yes, you'd need to create CronTab object manually. you can set metadata using ObjectMetaBuilder, however you'd need to populate spec manually. You can see my demo project for example https://github.com/rohanKanojia/kubernetes-client-demo/blob/master/src/main/java/io/fabric8/TypedCustomResourceDemoBlog.java

You're right about our documentation being outdated. Could you please create an issue about this :pray: ?

**Reply by TheCodeAddiction** [reply]
Sure! I can create an issue :) if you need help I could also try to get find some time and help you guys update it :)

**Reply by rohanKanojia** [reply]
that would be great.
